<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) 
FROM stores;

"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * 
FROM stores
ORDER BY name ASC
LIMIT  4;

"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

In [4]:
sql = """
SELECT * 
FROM stores
ORDER BY stores.name
LIMIT 4;
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

**Prompt 2 Answer**:

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [6]:

sql = """
SELECT * 
FROM products
"""
products = pd.read_sql_query(sql, engine)

In [10]:
products['item_description'] = 'VODKA'
products.columns

Index(['item_no', 'category_name', 'item_description', 'vendor', 'vendor_name',
       'bottle_size', 'pack', 'inner_pack', 'age', 'proof', 'list_date', 'upc',
       'scc', 'bottle_price', 'shelf_price', 'case_cost'],
      dtype='object')

In [5]:
#Among those sales whose category name is IMPORTED VODKA, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
#You should give us three sets of numbers - one for each transaction.

sql = """
SELECT btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor ASC
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty
0     $8.99          12
1    $28.50           1
2    $28.50           3

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [23]:
products['item_description'] = 'liquor'
products

item_no                       category_name item_description  vendor  \
0      904616  MISC. AMERICAN CORDIALS & LIQUEURS           liquor     305   
1      904617  MISC. IMPORTED CORDIALS & LIQUEURS           liquor     391   
2      904618                      PEACH BRANDIES           liquor     370   
3      904619                     SCOTCH WHISKIES           liquor     305   
4      904620                      FLAVORED VODKA           liquor     380   
...       ...                                 ...              ...     ...   
9972   904611                      CREAM LIQUEURS           liquor     109   
9973   904612                  SINGLE MALT SCOTCH           liquor     391   
9974   904613         DISTILLED SPIRITS SPECIALTY           liquor     384   
9975   904614  MISC. AMERICAN CORDIALS & LIQUEURS           liquor     194   
9976   904615                      FLAVORED VODKA           liquor     260   

                             vendor_name  bottle_size  pack  inner_pack   age  \
0                                Mhw Ltd          750    12           1  None   
1     Anchor Distilling (preiss Imports)          750    12           1  None   
2       Pernod Ricard Usa/austin Nichols         1000    12           1  None   
3                                Mhw Ltd          750     6           1  None   
4              Phillips Beverage Company          200    24           1  None   
...                                  ...          ...   ...         ...   ...   
9972                       Casa 1921 Llc          750     6           1  None   
9973  Anchor Distilling (preiss Imports)          750     1           1  None   
9974                 Piedmont Distillers          750     6           1  None   
9975                Fire Tail Brands Llc          750    12           1  None   
9976                     Diageo Americas          502    12          10  None   

     proof  list_date           upc             scc bottle_price  shelf_price  \
0       40 2009-02-11          None            None        $9.77        14.66   
1       60 2009-04-24          None            None       $14.12        21.18   
2       60 2009-12-16          None     89540500867        $6.50         9.74   
3       86 2009-12-24          None            None       $25.33        38.00   
4       60 2009-11-05          None  10087100000000        $1.97         2.96   
...    ...        ...           ...             ...          ...          ...   
9972    30 2009-12-08          None  10608600000000       $18.84        28.26   
9973    92 2009-12-09          None          904612       $63.00        94.50   
9974    80 2009-12-15  851767000000  10851800000000       $13.56        20.34   
9975    60 2010-02-05         90461          904614        $5.06         7.59   
9976    70 2009-12-04   82000768142            None        $5.72         8.58   

      case_cost  
0        117.22  
1        169.40  
2         77.94  
3        152.00  
4         47.30  
...         ...  
9972     113.00  
9973      63.00  
9974      81.35  
9975      60.70  
9976      68.58  

[9977 rows x 16 columns]

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(sales.total)
FROM sales
WHERE sales."date" BETWEEN '01-31-2015' AND '03-01-2015';
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**:

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.vendor, COUNT(*) AS transactions, SUM(sales.bottle_qty) AS bottle_qty, SUM(sales.total) AS total
FROM sales
WHERE sales.category_name = 'IMPORTED VODKA'
GROUP BY sales.vendor
ORDER BY COUNT(*) DESC;
"""

df = pd.read_sql_query(sql, engine)
df.head()

vendor  transactions  bottle_qty       total
0  Pernod Ricard USA/Austin Nichols         42338      403450  7880365.88
1               Bacardi U.S.A. Inc.         26441      236047  5771242.07
2                   Diageo Americas         15725      152038  3346425.42
3   Constellation Wine Company Inc.         12565      116053  1895005.51
4                   Jim Beam Brands          5850      104119  1807097.23

**Prompt 5 Answer**: 
> Pernod Ricard USA/Austin Nichols

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost as Markup
FROM sales
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [9]:
# Use a SQL query to find the answer to the above prompt.
sql = """
SELECT store, AVG((btl_price - state_btl_cost) / state_btl_cost) AS "Percentage Markup"
FROM sales
GROUP BY store
ORDER BY "Percentage Markup" DESC;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup
0      4013           0.525471
1      4024           0.521499
2      4266           0.518903
3      3833           0.515094
4      3660           0.514591
...     ...                ...
1347   9018           0.500000
1348   4784           0.499994
1349   4494           0.499909
1350   4534           0.499781
1351   4693           0.499214

[1352 rows x 2 columns]

**Prompt 7 Answer**: 
> The store with the highest average markup percentage is store 4013, which has an average percentage markup of 52.5471%.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales.store, AVG((sales.btl_price - sales.state_btl_cost) / sales.state_btl_cost) AS "Percentage Markup", stores.name, stores.store_address
FROM sales 
    LEFT JOIN stores
    ON sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY "Percentage Markup" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup                        name  \
0   4013           0.525471      Wal-Mart 0841 / Tipton   
1   4024           0.521499  Wal-Mart 1546 / Iowa Falls   
2   4266           0.518903  Wal-Mart 1683 / Shenandoah   
3   3833           0.515094    Wal-Mart 3394 / Atlantic   
4   3660           0.514591   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**:
- Wal-Mart 0841
- Wal-Mart 1546 
- Wal-Mart 1683
- Wal-Mart 3394
- Wal-Mart 2935

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT counties.county, (SUM(bottle_qty) / counties.population) AS "Bottles Sold per Capita", ROUND(AVG(products.bottle_size),1) AS "Average Bottle Size"
FROM counties
    LEFT JOIN sales
        LEFT JOIN products
        ON sales.item = products.item_no
    ON counties.county = sales.county
GROUP BY counties.county
ORDER BY "Bottles Sold per Capita" DESC
LIMIT 10;
"""

pd.read_sql_query(sql, engine)

county  Bottles Sold per Capita  Average Bottle Size
0      Dickinson                       22                974.1
1           Polk                       15                892.6
2     Black Hawk                       15                862.9
3    Cerro Gordo                       14                986.4
4        Johnson                       13                895.4
5          Scott                       13                871.6
6           Linn                       12                886.3
7  Pottawattamie                       11                893.7
8        Kossuth                       11               1041.0
9            Lee                       10                912.2

**Prompt 9 Answer**:
>Kossuth County sells the largest average bottle size (among the ten counties selling the most liquor bottles per capita), with an average bottle size of 1041.0 milliliters.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [12]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(*)
FROM sales 
WHERE total::money != btl_price * bottle_qty
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**:
> Have no records

1.How many total products are in the Products table?



In [15]:
sql = """
SELECT COUNT(*) 
FROM products;

"""

pd.read_sql_query(sql, engine)

count
0   9977

2.Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? How many 
different products do they have?


3.Which products sell the best by total number of unit sales? (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)



In [17]:
sql = """
SELECT * 
FROM products;

"""

pd.read_sql_query(sql, engine)

item_no                       category_name  \
0      904616  MISC. AMERICAN CORDIALS & LIQUEURS   
1      904617  MISC. IMPORTED CORDIALS & LIQUEURS   
2      904618                      PEACH BRANDIES   
3      904619                     SCOTCH WHISKIES   
4      904620                      FLAVORED VODKA   
...       ...                                 ...   
9972   904611                      CREAM LIQUEURS   
9973   904612                  SINGLE MALT SCOTCH   
9974   904613         DISTILLED SPIRITS SPECIALTY   
9975   904614  MISC. AMERICAN CORDIALS & LIQUEURS   
9976   904615                      FLAVORED VODKA   

                                  item_description  vendor  \
0                           Travis Hasse Apple Pie     305   
1                               D'aristi Xtabentun     391   
2                        Hiram Walker Peach Brandy     370   
3                                 Oak Cross Whisky     305   
4                             Uv Red(cherry) Vodka     380   
...                                            ...     ...   
9972                            1921 Tequila Cream     109   
9973                              Glendronach 15yo     391   
9974  Junior Johnson's Midnight Moon Lemonade(dno)     384   
9975                          Jekyll Berry Liqueur     194   
9976                 Smirnoff Twist Of Citrus Mini     260   

                             vendor_name  bottle_size  pack  inner_pack   age  \
0                                Mhw Ltd          750    12           1  None   
1     Anchor Distilling (preiss Imports)          750    12           1  None   
2       Pernod Ricard Usa/austin Nichols         1000    12           1  None   
3                                Mhw Ltd          750     6           1  None   
4              Phillips Beverage Company          200    24           1  None   
...                                  ...          ...   ...         ...   ...   
9972                       Casa 1921 Llc          750     6           1  None   
9973  Anchor Distilling (preiss Imports)          750     1           1  None   
9974                 Piedmont Distillers          750     6           1  None   
9975                Fire Tail Brands Llc          750    12           1  None   
9976                     Diageo Americas          502    12          10  None   

     proof  list_date           upc             scc bottle_price  shelf_price  \
0       40 2009-02-11          None            None        $9.77        14.66   
1       60 2009-04-24          None            None       $14.12        21.18   
2       60 2009-12-16          None     89540500867        $6.50         9.74   
3       86 2009-12-24          None            None       $25.33        38.00   
4       60 2009-11-05          None  10087100000000        $1.97         2.96   
...    ...        ...           ...             ...          ...          ...   
9972    30 2009-12-08          None  10608600000000       $18.84        28.26   
9973    92 2009-12-09          None          904612       $63.00        94.50   
9974    80 2009-12-15  851767000000  10851800000000       $13.56        20.34   
9975    60 2010-02-05         90461          904614        $5.06         7.59   
9976    70 2009-12-04   82000768142            None        $5.72         8.58   

      case_cost  
0        117.22  
1        169.40  
2         77.94  
3        152.00  
4         47.30  
...         ...  
9972     113.00  
9973      63.00  
9974      81.35  
9975      60.70  
9976      68.58  

[9977 rows x 16 columns]

4.Which products sell the best by total dollar value of sales?
    


5.What are the top 10 categories of liquor sold based on the total amount of sales revenue?



6.Which rum products have sales greater than $10,000? How about whiskey or vodka products?



7.Which county sold the most amount of vodka during February 2014?



8.Which counties were in the top 10 counties for vodka sales in any month in 2014?



9.Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” list over the course of 2014.



10.What is the trend of sales by month? Break up variables such as bottle_price or liter_size into categories (for example: cheap, medium, or expensive).



11.Which stores sell one of the top five most expensive bottles of alcohol?



12.How many stores have more than $2,000,000 in total sales?



13.How many stores have an average bottle price greater than $20?



14.Which stores have the highest sales of items over 90 proof? (edited) 





15.Store 2238 (Adventureland Inn at 3200 Adventureland Dr) sold $883.24 in April and $27,526.38 in May, for a 3017% growth rate. That was the highest percentage month-on-month growth rate. Create a query that shows this and the next 9 highest after that. (edited) 





16.We think the data might have been corrupted in some way. The category listed in the sales table doesn’t always match up with the category in the products table. How many times has this happened, and can you find any patterns to it?



17.The store_address field in the stores table actually contains three rows of text. Quite often the latitude and longitude are in the last line of text. Create a query that shows the geo-locatable stores in latitude order (i.e. show the stores from the most northerly to the most southerly).